tasks/change-granule-collection-s3/src/index.ts

```diff
@@ -0,0 +1,560 @@
+'use strict';
+
+import { Context } from 'aws-lambda';
+import get from 'lodash/get';
+import keyBy from 'lodash/keyBy';
+import cloneDeep from 'lodash/cloneDeep';
+import noop from 'lodash/noop';
+import { AssertionError } from 'assert';
+import flatten from 'lodash/flatten';
+import pRetry from 'p-retry';
+import path from 'path';
+import pMap from 'p-map';
+
+import { InvalidArgument, DuplicateFile, ValidationError } from '@cumulus/errors';
+import {
+  unversionFilename,
+} from '@cumulus/ingest/granule';
+import {
+  isCMRFile,
+  metadataObjectFromCMRFile,
+} from '@cumulus/cmrjs';
+import { runCumulusTask } from '@cumulus/cumulus-message-adapter-js';
+import { s3 } from '@cumulus/aws-client/services';
+import { BucketsConfig } from '@cumulus/common';
+import { urlPathTemplate } from '@cumulus/ingest/url-path-template';
+import { constructCollectionId } from '@cumulus/message/Collections';
+import { getCollection } from '@cumulus/api-client/collections';
+import { getGranule } from '@cumulus/api-client/granules';
+import { CollectionRecord, CollectionFile } from '@cumulus/types';
+import { CumulusMessage } from '@cumulus/types/message';
+import { getRequiredEnvVar } from '@cumulus/common/env';
+import { log } from '@cumulus/common';
+import { calculateObjectHash, copyObject, s3Join, s3ObjectExists } from '@cumulus/aws-client/S3';
+import { fetchDistributionBucketMap } from '@cumulus/distribution-utils';
+import { getCMRCollectionId } from '@cumulus/cmrjs/cmr-utils';
+import {
+  MB,
+  EventConfig,
+  ChangeCollectionsS3Event,
+  ValidApiGranuleFile,
+  ValidGranuleRecord,
+  MassagedEventConfig,
+  ValidApiFile,
+} from './types';
+import {
+  apiFileIsValid,
+  apiGranuleRecordIsValid,
+  CMRObjectToString,
+  isCMRMetadataFile,
+  updateCmrFileCollections,
+  uploadCMRFile,
+} from './update_cmr_file_collection';
+
+/**
+ * Boolean are these files in the same location
+ */
+function objectSourceAndTargetSame(
+  sourceFile: ValidApiGranuleFile,
+  targetFile: ValidApiGranuleFile
+): boolean {
+  return ((sourceFile.key === targetFile.key) && (sourceFile.bucket === targetFile.bucket));
+}
+
+async function metadataCollisionsMatch(
+  targetFile: ValidApiGranuleFile,
+  metadataObject: Object
+): Promise<boolean> {
+  const existingGranuleMetadata = await pRetry(
+    () => metadataObjectFromCMRFile(
+      `s3://${targetFile.bucket}/${targetFile.key}`
+    ),
+    { retries: 5, minTimeout: 2000, maxTimeout: 2000 }
+  );
+  const sourceCollection = getCMRCollectionId(metadataObject, targetFile.key);
```

Comment:
Is it valid here that both of these could return as a match if the CMR metadata doesn't have a collection ID (get in getCMRCollectionID will return undefined, which will match)? 

Suggest at a minimum we log and warn that condition, but noting for consideration tomorrow.